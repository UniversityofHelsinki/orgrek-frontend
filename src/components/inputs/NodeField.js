import React, { useState } from 'react';
import PropTypes from 'prop-types';
import TextField from '@mui/material/TextField';
import Autocomplete from '@mui/material/Autocomplete';
import { useTranslation } from 'react-i18next';
import useTree from '../../hooks/useTree';
import useContentLanguage from '../../hooks/useContentLanguage';
import InputAdornment from '@mui/material/InputAdornment';
import SearchIcon from '../icons/Search';

const flatten = (current) =>
  current.reduce((a, c) => [...a, c, ...flatten(c.children)], []);

const nameMatches = (name, text) => {
  return name.toLowerCase().indexOf(text.toLowerCase()) !== -1;
};

const uniqueIdMatches = (uniqueId, text) => {
  return uniqueId.toString() === text.toLowerCase();
};

/**
 * Autocomplete field for searching and selecting a node from the tree.
 *
 * See MUI Autocomplete for all supported props.
 */
const NodeField = ({
  label,
  placeholder,
  helperText,
  variant,
  disabled,
  error,
  required,
  value,
  onChange,
  clearOnSelect = false,
  ...props
}) => {
  const { t } = useTranslation();
  const { tree } = useTree();
  const language = useContentLanguage();
  const [inputValue, setInputValue] = useState('');

  const options = tree && tree[language] ? flatten([tree[language]]) : [];

  const uniqueOptions = [
    ...new Map(options.map((item) => [item['uniqueId'], item])).values(),
  ];

  const handleChange = (event, newValue, reason) => {
    onChange(
      event,
      newValue === null ? null : { id: newValue.uniqueId, name: newValue.name },
      reason
    );
    if (clearOnSelect) {
      setTimeout(() => setInputValue(''));
    }
  };

  let inputAdornment;
  if (variant === 'search') {
    inputAdornment = (
      <InputAdornment position="end">
        <SearchIcon color={disabled ? 'disabled' : 'primary'} />
      </InputAdornment>
    );
  }

  return (
    <Autocomplete
      clearText={t('nodeField.clearText')}
      openText={t('nodeField.openText')}
      closeText={t('nodeField.closeText')}
      noOptionsText={t('nodeField.noOptionsText')}
      {...props}
      disabled={disabled}
      freeSolo={variant === 'search'}
      onChange={handleChange}
      value={value}
      isOptionEqualToValue={(option, value) => {
        return option.uniqueId === value.id;
      }}
      inputValue={inputValue}
      onInputChange={(event, value) => setInputValue(value)}
      options={uniqueOptions}
      getOptionLabel={(option) => option.name || ''}
      renderOption={(props, option) => (
        <li {...props} key={`${option.name}`}>
          {option.name}
        </li>
      )}
      filterOptions={(options, state) => {
        if (state.inputValue.length > 2) {
          return options.filter(
            (option) =>
              nameMatches(option.name, state.inputValue) ||
              uniqueIdMatches(option.uniqueId, state.inputValue)
          );
        }
        return [];
      }}
      renderInput={({ InputProps, ...params }) => (
        <TextField
          InputProps={{
            ...InputProps,
            endAdornment: InputProps.endAdornment || inputAdornment,
          }}
          label={label}
          placeholder={placeholder || t('type_three_char_to_start_search')}
          helperText={helperText}
          required={required}
          error={error}
          {...params}
        />
      )}
    />
  );
};

NodeField.propTypes = {
  /** Field label */
  label: PropTypes.string,

  /** Hint text displayed in the input before the user enters a value. */
  placeholder: PropTypes.string,

  /**
   * Variant 'combobox' allows only selecting a value from the suggestion menu,
   * and the field is cleared on blur if nothing was selected.
   *
   * Variant 'search' keeps the typed text in the field on blur even if it is
   * not a valid selection.
   * In addition, displays a search icon as input adornment.
   */
  variant: PropTypes.oneOf(['combobox', 'search']),

  /**
   * Called when a value is selected or the field is cleared.
   */
  onChange: PropTypes.func.isRequired,

  /**
   * Helper text displayed below the field.
   *
   * HelperText component can be used with this prop to display errors.
   */
  helperText: PropTypes.oneOfType([PropTypes.string, PropTypes.element]),

  /** If true, the component is disabled. */
  disabled: PropTypes.bool,

  /** If true, the field is displayed in an error state. */
  error: PropTypes.bool,

  /** If true, the label is displayed as required and the input value is required. */
  required: PropTypes.bool,

  /** Field value as object containing node unique id and name */
  value: PropTypes.shape({
    id: PropTypes.number.isRequired,
    name: PropTypes.string.isRequired,
  }),
};

export default NodeField;
