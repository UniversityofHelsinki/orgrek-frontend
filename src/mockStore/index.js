export * from './mockApi';
export * from './mockDay';
export * from './mockEditMode';
export * from './mockHierarchy';
export * from './mockHierarchyFilters';
export * from './mockNode';
export * from './mockNodeDetailsView';
export * from './mockStore';
export * from './mockTree';
export * from './mockUser';
