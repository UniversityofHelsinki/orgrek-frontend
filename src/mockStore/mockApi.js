import { rest } from 'msw';
import format from 'date-fns/format';

const formatApiDate = (date) => format(date, 'd.M.yyyy');

const baseUrl = `${process.env.REACT_APP_ORGREK_BACKEND_SERVER || ''}/api`;

export const mockGetNode = (nodeId, body) =>
  rest.get(`${baseUrl}/node/${nodeId}`, (req, res, ctx) => res(ctx.json(body)));

export const mockGetAttributes = (nodeId, body, date) => {
  return rest.get(
    `${baseUrl}/node/${nodeId}/virallinen/${formatApiDate(
      date || new Date()
    )}/attributes`,
    (req, res, ctx) => res(ctx.json(body))
  );
};

export const mockPutNameAttributes = (nodeId) =>
  rest.put(`${baseUrl}/node/${nodeId}/attributes/names`, (req, res, ctx) =>
    res(ctx.delay(2000), ctx.json(req.json()))
  );

export const mockGetTypeAttributes = (nodeId, body) =>
  rest.get(`${baseUrl}/node/${nodeId}/attributes/types`, (req, res, ctx) =>
    res(ctx.json(body))
  );

export const mockPutTypeAttributes = (nodeId) =>
  rest.put(`${baseUrl}/node/${nodeId}/attributes/types`, (req, res, ctx) =>
    res(ctx.delay(2000), ctx.json(req.json()))
  );

export const mockGetValidHierarchyFilters = (date, body) =>
  rest.get(
    `${baseUrl}/hierarchyFilters/${formatApiDate(date)}`,
    (req, res, ctx) => res(ctx.json(body))
  );

export const mockGetCodeAttributes = (nodeId, body, date) =>
  rest.get(`${baseUrl}/node/${nodeId}/attributes/codes`, (req, res, ctx) =>
    res(ctx.json(body))
  );

export const mockSaveCodeAttributes = (nodeId) =>
  rest.put(`${baseUrl}/node/${nodeId}/attributes/codes`, (req, res, ctx) =>
    res(ctx.delay(2000), ctx.json(req.json()))
  );

export const mockGetOtherAttributes = (nodeId, selectedHierarchies, body) =>
  rest.get(
    `${baseUrl}/node/${nodeId}/attributes/others/hierarchies/${selectedHierarchies}`,
    (req, res, ctx) => res(ctx.json(body))
  );

export const mockSaveOtherAttributes = (nodeId) =>
  rest.put(`${baseUrl}/node/${nodeId}/attributes/others`, (req, res, ctx) =>
    res(ctx.delay(2000), ctx.json(req.json()))
  );

export const mockGetAttributeKeys = ({ selectedHierarchies, sections }, body) =>
  rest.get(
    `${baseUrl}/hierarchyFilters/${selectedHierarchies}/${
      (sections && sections.toString()) || 'virallinen'
    }/attributes/keys`,
    (req, res, ctx) => res(ctx.json(body))
  );

export const mockGetFullNames = (nodeId, date, body) =>
  rest.get(
    `${baseUrl}/node/fullname/${nodeId}/${formatApiDate(date)}`,
    (req, res, ctx) => res(ctx.json(body))
  );

export const mockGetAllFullNames = (nodeId, body) =>
  rest.get(`${baseUrl}/node/fullname/all/${nodeId}`, (req, res, ctx) =>
    res(ctx.json(body))
  );

export const mockGetFavorableFullNames = (nodeId, date, body) =>
  rest.get(
    `${baseUrl}/node/fullname/favorable/${nodeId}/${formatApiDate(date)}`,
    (req, res, ctx) => res(ctx.json(body))
  );

export const mockGetParents = (nodeId, date, hierarchies, body) =>
  rest.get(
    `${baseUrl}/node/${nodeId}/${date.toLocaleDateString(
      'fi-FI'
    )}/parents/${hierarchies}`,
    (req, res, ctx) => res(ctx.json(body))
  );

export const mockSaveParents = () =>
  rest.put(`${baseUrl}/node/parents/update`, (req, res, ctx) =>
    res(ctx.delay(2000), ctx.json(req.json()))
  );

export const mockGetChildren = (nodeId, date, hierarchies, body) =>
  rest.get(
    `${baseUrl}/node/${nodeId}/${date.toLocaleDateString(
      'fi-FI'
    )}/children/${hierarchies}`,
    (req, res, ctx) => res(ctx.json(body))
  );

export const mockSaveChildren = () =>
  rest.put(`${baseUrl}/node/children/update`, (req, res, ctx) =>
    res(ctx.delay(2000), ctx.json(req.json()))
  );

export const mockGetPredecessors = (nodeId, date, body) =>
  rest.get(
    `${baseUrl}/node/${nodeId}/${date.toLocaleDateString(
      'fi-FI'
    )}/predecessors`,
    (req, res, ctx) => res(ctx.json(body))
  );

export const mockGetSuccessors = (nodeId, date, body) =>
  rest.get(
    `${baseUrl}/node/${nodeId}/${date.toLocaleDateString('fi-FI')}/successors`,
    (req, res, ctx) => res(ctx.json(body))
  );

export const mockSaveSuccessors = (nodeId) =>
  rest.put(`${baseUrl}/node/${nodeId}/successors`, (req, res, ctx) =>
    res(ctx.delay(2000), ctx.json(req.json()))
  );

export const mockGetSectionTypeAttributes = (sectionType, body) =>
  rest.get(
    `${baseUrl}/node/section/${sectionType}/attributes`,
    (req, res, ctx) => res(ctx.json(body))
  );

export const mockGetTree = ({ hierarchies, selectedDay }, body) =>
  rest.get(
    `${baseUrl}/tree/${hierarchies}/${selectedDay.toLocaleDateString('FI-fi')}`,
    (req, res, ctx) => res(ctx.json(body))
  );

export const mockGetNodeValidity = (id, body) =>
  rest.get(`${baseUrl}/node/${id}`, (req, res, ctx) => res(ctx.json(body)));

export const mockSaveNodeValidity = (id) =>
  rest.put(`${baseUrl}/node/${id}/update`, (req, res, ctx) =>
    res(ctx.delay(2000), ctx.json(req.json()))
  );

/**
 * @deprecated not needed after everything has been migrated to RTK Query
 */
export const mockGetParentsDeprecated = (nodeId, date, hierarchies, body) =>
  rest.get(
    `${baseUrl}/node/parents/${nodeId}/${formatApiDate(date)}/${hierarchies}`,
    (req, res, ctx) => res(ctx.json(body))
  );
