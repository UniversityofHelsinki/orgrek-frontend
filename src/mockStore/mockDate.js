import React from 'react';

const RealDate = Date;

/**
 * Storybook decorator for mocking current time.
 *
 * Sets current time to a fixed value if systemTime is defined in story
 * parameters, otherwise uses the real time.
 */
export const withMockDate = (Story, { parameters }) => {
  if (parameters.systemTime instanceof Date) {
    // eslint-disable-next-line no-global-assign
    Date = class extends Date {
      constructor(...args) {
        if (args.length === 0) {
          super(parameters.systemTime.valueOf());
        } else {
          super(...args);
        }
      }

      static now() {
        return parameters.systemTime.getTime();
      }
    };
  } else {
    // eslint-disable-next-line no-global-assign
    Date = RealDate;
  }

  return <Story />;
};
