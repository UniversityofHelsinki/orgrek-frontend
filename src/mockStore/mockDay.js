import { createSlice } from '@reduxjs/toolkit';
import legacyReducer from '../reducers/dayChangeReducer';

const initialState = {
  selectedDay: new Date(),
};

const daySlice = createSlice({
  name: 'day',
  initialState,
  extraReducers(builder) {
    builder.addDefaultCase(legacyReducer);
  },
});

export const dayReducer = daySlice.reducer;
