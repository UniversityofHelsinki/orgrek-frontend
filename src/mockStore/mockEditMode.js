import { createSlice } from '@reduxjs/toolkit';
import legacyReducer from '../reducers/editModeReducer';

const initialState = {
  edit: false,
};

/**
 * @deprecated used only in the legacy edit view
 */
const editModeSlice = createSlice({
  name: 'editMode',
  initialState,
  extraReducers(builder) {
    builder.addDefaultCase(legacyReducer);
  },
});

/**
 * @deprecated used only in the legacy edit view
 */
export const editModeReducer = editModeSlice.reducer;
