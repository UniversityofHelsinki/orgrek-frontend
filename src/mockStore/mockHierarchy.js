const defaultHierarchies = [
  'tutkimus',
  'henkilosto',
  'toiminnanohjaus',
  'opetus',
  'history',
  'talous',
  'virallinen',
];

export const createHierarchies = (
  hierarchies = null,
  startDate = null,
  endDate = null
) =>
  (hierarchies || defaultHierarchies).map((hierarchy) => ({
    hierarchy,
    startDate,
    endDate,
  }));
