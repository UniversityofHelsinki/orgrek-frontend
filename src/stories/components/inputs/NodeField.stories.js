import { expect } from '@storybook/jest';
import NodeField from '../../../components/inputs/NodeField';
import { mockGetTree, withMockStore, tree } from '../../../mockStore';
import { userEvent, within, waitFor } from '@storybook/testing-library';

// Use a fixed date to ensure that tests always have a consistent result
const now = new Date('2023-03-22T14:28:00+0200');

const selectedHierarchy = 'talous';

export default {
  component: NodeField,
  parameters: {
    systemTime: now,
  },
};

export const Combobox = {
  args: {
    variant: 'combobox',
    label: 'Yksikkö',
    placeholder: '',
    disabled: false,
  },
  parameters: {
    msw: {
      handlers: [
        mockGetTree({ hierarchies: selectedHierarchy, selectedDay: now }, tree),
      ],
    },
  },
  decorators: [
    withMockStore({
      dr: {
        selectedDay: now,
      },
      tree: {
        selectedHierarchy,
      },
    }),
  ],
};

export const Search = {
  ...Combobox,
  args: {
    ...Combobox.args,
    label: 'Hae yksikköä',
    variant: 'search',
  },
};

export const Selected = {
  ...Combobox,
  play: async ({ canvasElement }) => {
    const canvas = within(canvasElement.parentElement);

    await userEvent.type(canvas.getByLabelText('Yksikkö'), 'kie');

    await waitFor(async () => {
      await expect(canvas.getByText(/HY, Kielikeskus/)).toBeInTheDocument();
    });

    await userEvent.click(canvas.getByText(/HY, Kielikeskus/));
  },
};
