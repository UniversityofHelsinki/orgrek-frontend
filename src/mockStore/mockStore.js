import { configureStore } from '@reduxjs/toolkit';
import { treeReducer } from './mockTree';
import { hierarchyReducer } from './mockHierarchy';
import { createNodeState, nodeReducer } from './mockNode';
import { userReducer } from './mockUser';
import { dayReducer } from './mockDay';
import { nodeDetailsViewReducer } from './mockNodeDetailsView';
import { hierarchyFiltersReducer } from './mockHierarchyFilters';
import { editModeReducer } from './mockEditMode';
import { api, notificationsReducer } from '../store';
import { Provider } from 'react-redux';
import React from 'react';
import { setupListeners } from '@reduxjs/toolkit/query';

export const configureMockStore = (preloadedState = {}) => {
  const store = configureStore({
    preloadedState,
    reducer: {
      [api.reducerPath]: api.reducer,
      tree: treeReducer,
      hr: hierarchyReducer,
      nrd: nodeReducer,
      ur: userReducer,
      dr: dayReducer,
      nvrd: nodeDetailsViewReducer,
      hierarchyFilters: hierarchyFiltersReducer,
      editModeReducer: editModeReducer,
      notifications: notificationsReducer,
    },
    middleware: (getDefaultMiddleware) =>
      getDefaultMiddleware().concat(api.middleware),
  });

  setupListeners(store.dispatch);

  return store;
};

/**
 * Configures Redux store from initial state and any additional test data
 * passed to this component.
 *
 * Intended to be used as Storybook story decorator, but it should work in
 * tests as well.
 *
 * @param state
 * @param children a Storybook story or any component using the Redux store
 */
export const MockStoreProvider = ({ state, children }) => {
  const store = configureMockStore(state);

  return <Provider store={store}>{children}</Provider>;
};

/**
 * Storybook decorator function which provides the mock Redux store to the story
 *
 * @param state
 * @return a function that can be passed to Storybook decorator array
 */
export const withMockStore = (state) => {
  return function MockStoreDecorator(Story) {
    return (
      <MockStoreProvider state={state}>
        <Story />
      </MockStoreProvider>
    );
  };
};

/**
 * For convenience a decorator that mocks Redux store with the given user.
 *
 * Use this decorator when you need to mock only the current user.
 * Alternatively, withMockStore decorator can mock the whole state.
 *
 * @param user current user
 */
export const withUser = (user) => withMockStore({ ur: { user } });

/**
 * For convenience a decorator that mocks Redux store with the given node data.
 *
 * Use this decorator when you need to mock only the node reducer state.
 * Alternatively, withMockStore decorator can mock the whole state.
 *
 * @param nodeState overrides node reducer initial state
 */
export const withNode = (nodeState) =>
  withMockStore({ nrd: createNodeState(nodeState) });

/**
 * For convenience a decorator that mocks Redux store with the given hierarchy
 * data.
 *
 * Use this decorator when you need to mock only the hierarchy reducer state.
 * Alternatively, withMockStore decorator can mock the whole state.
 *
 * @param hierarchyState overrides hierarchy reducer initial state
 */
export const withHierarchy = (hierarchyState) =>
  withMockStore({ hr: { ...hierarchyState } });
