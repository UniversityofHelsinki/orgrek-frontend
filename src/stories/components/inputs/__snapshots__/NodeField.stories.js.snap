// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`stories/components/inputs/NodeField Combobox smoke-test 1`] = `
Object {
  "checked": undefined,
  "children": Array [
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "Yksikkö",
      "pressed": undefined,
      "role": "text",
      "value": undefined,
    },
    Object {
      "autocomplete": "list",
      "checked": undefined,
      "children": undefined,
      "haspopup": "listbox",
      "name": "Yksikkö",
      "pressed": undefined,
      "role": "combobox",
      "value": undefined,
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "description": "Avaa",
      "name": "Avaa",
      "pressed": undefined,
      "role": "button",
      "value": undefined,
    },
  ],
  "name": "stories-components-inputs-nodefield--combobox",
  "pressed": undefined,
  "role": "WebArea",
  "value": undefined,
}
`;

exports[`stories/components/inputs/NodeField Search smoke-test 1`] = `
Object {
  "checked": undefined,
  "children": Array [
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "Hae yksikköä",
      "pressed": undefined,
      "role": "text",
      "value": undefined,
    },
    Object {
      "autocomplete": "list",
      "checked": undefined,
      "children": undefined,
      "haspopup": "listbox",
      "name": "Hae yksikköä",
      "pressed": undefined,
      "role": "combobox",
      "value": undefined,
    },
  ],
  "name": "stories-components-inputs-nodefield--search",
  "pressed": undefined,
  "role": "WebArea",
  "value": undefined,
}
`;

exports[`stories/components/inputs/NodeField Selected play-test 1`] = `
Object {
  "checked": undefined,
  "children": Array [
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "Yksikkö",
      "pressed": undefined,
      "role": "text",
      "value": undefined,
    },
    Object {
      "autocomplete": "list",
      "checked": undefined,
      "children": undefined,
      "focused": true,
      "haspopup": "listbox",
      "name": "Yksikkö",
      "pressed": undefined,
      "role": "combobox",
      "value": "HY, Kielikeskus (KIELIKESKUS)",
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "description": "Tyhjennä",
      "name": "Tyhjennä",
      "pressed": undefined,
      "role": "button",
      "value": undefined,
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "description": "Avaa",
      "name": "Avaa",
      "pressed": undefined,
      "role": "button",
      "value": undefined,
    },
  ],
  "name": "stories-components-inputs-nodefield--selected",
  "pressed": undefined,
  "role": "WebArea",
  "value": undefined,
}
`;
